rust/agents/relayer/src/msg/metadata/base.rs

```diff
@@ -136,8 +157,15 @@ impl BaseMetadataBuilder {
         }
     }
 
-    pub async fn highest_known_nonce(&self) -> u32 {
-        self.origin_prover_sync.read().await.count() - 1
+    pub async fn highest_known_nonce(&self) -> Option<u32> {
+        self.origin_prover_sync.read().await.count().checked_sub(1)
+    }
+
+    pub async fn get_merkle_leaf_by_message_id(&self, message_id: H256) -> Result<Option<u32>> {
```

Comment:
nit: imo we should clarify this is getting the index, not the leaf itself, ie rename to `get_merkle_leaf_index_by_message_id`